pkg/nodejs/yarn/parse.go

```diff
@@ -60,69 +114,189 @@ func validProtocol(protocol string) (valid bool) {
+		return i + 2, data[0:i], nil
+	} else if i := bytes.Index(data, []byte("\r\n\r\n")); i >= 0 {
+		return i + 4, data[0:i], nil
+	}
+
+	// If we're at EOF, we have a final, non-terminated line. Return it.
+	if atEOF {
+		return len(data), data, nil
+	}
+	// Request more data.
+	return 0, nil, nil
+}
+
+func parseBlock(block []byte, lineNum int) (lib Library, deps []Dependency, newLine int, err error) {
+	var (
+		emptyLines int // lib can start with empty lines first
+		skipBlock  bool
+	)
+
+	scanner := NewLineScanner(bytes.NewReader(block))
 	for scanner.Scan() {
-		lineNumber++
 		line := scanner.Text()
-		if len(line) < 1 {
+
+		if len(line) == 0 {
+			emptyLines++
 			continue
 		}
 
-		// parse version
-		var version string
-		if version, err = getVersion(line); err == nil {
-			if skipPackage {
-				continue
-			}
-			if lib.Name == "" {
-				return nil, nil, xerrors.New("Invalid yarn.lock format")
-			}
-			// fetch between version prefix and last double-quote
-			symbol := fmt.Sprintf("%s@%s", lib.Name, version)
-			if _, ok := unique[symbol]; ok {
-				lib = types.Library{}
-				continue
-			}
+		if line[0] == '#' || skipBlock {
+			continue
+		}
 
-			lib.Version = version
-			libs = append(libs, lib)
-			lib = types.Library{}
-			unique[symbol] = struct{}{}
+		// Skip this block
+		if strings.HasPrefix(line, "__metadata") {
+			skipBlock = true
 			continue
 		}
-		// skip __metadata block
-		if skipPackage = strings.HasPrefix(line, "__metadata"); skipPackage {
+
+		line = strings.TrimPrefix(strings.TrimSpace(line), "\"")
+
+		switch {
+		case strings.HasPrefix(line, "version"):
+			if lib.Version, err = getVersion(line); err != nil {
+				skipBlock = true
+			}
+			continue
+		case strings.HasPrefix(line, "dependencies:"):
+			// start dependencies block
+			deps = parseDependencies(scanner)
 			continue
 		}
-		// packagename line start 1 char
-		if line[:1] != " " && line[:1] != "#" {
-			var name string
-			var protocol string
-			if name, protocol, err = parsePackageLocator(line); err != nil {
+
+		// try parse package locator
+		if name, protocol, locs, locErr := parsePackageLocators(line); locErr == nil {
+			if locs == nil || !validProtocol(protocol) {
+				skipBlock = true
+				err = xerrors.Errorf("failed to parse package locator")
+				continue
+			} else {
+				lib.Locators = locs
+				lib.Name = name
 				continue
 			}
-			if skipPackage = !validProtocol(protocol); skipPackage {
+		}
+	}
+
+	lib.Location = types.Location{
+		StartLine: lineNum + emptyLines,
+		EndLine:   scanner.LineNum(lineNum),
+	}
+
+	if scanErr := scanner.Err(); err != scanErr {
+		err = scanErr
+	}
+
+	return lib, deps, scanner.LineNum(lineNum), err
+}
+
+func parseDependencies(scanner *LineScanner) (deps []Dependency) {
+	for scanner.Scan() {
+		line := scanner.Text()
+		if dep, err := parseDependency(line); err != nil {
+			// finished dependencies block
+			return deps
+		} else {
+			deps = append(deps, dep)
+		}
```

Comment:
The only error it can get is an expected custom error, so no real need for this.
I can change it to a boolean or use an empty string to make it clearer.